ic-asset/src/asset_config.rs

```diff
@@ -0,0 +1,627 @@
+}
+
+impl AssetConfig {
+    fn merge(mut self, other: Self) -> Self {
+        if let Some(c) = other.cache {
+            self.cache = Some(c);
+        };
+        match (self.headers.as_mut(), other.headers) {
+            (Some(sh), Some(oh)) => sh.extend(oh),
+            (_, oh) => self.headers = oh,
+        };
+        self
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct AssetSourceDirectoryConfiguration {
+    config_map: Map,
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize)]
+pub(crate) struct AssetConfig {
+    pub(crate) cache: Option<CacheConfig>,
+    pub(crate) headers: Option<HeadersConfig>,
+}
+
+impl Default for AssetConfig {
+    fn default() -> Self {
+        Self {
+            headers: Some(HashMap::new()),
+            cache: Some(CacheConfig::default()),
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+struct AssetConfigTreeNode {
+    pub parent: Option<Arc<AssetConfigTreeNode>>,
+    pub rules: Vec<AssetConfigRule>,
+}
+
+impl AssetSourceDirectoryConfiguration {
+    pub(crate) fn load(root_dir: &Path) -> anyhow::Result<Self> {
+        let mut config_map = HashMap::new();
+        AssetConfigTreeNode::load(None, &root_dir, &mut config_map)?;
+        Ok(Self { config_map })
+    }
+
+    pub(crate) fn get_asset_config(&self, canonical_path: &Path) -> AssetConfig {
+        self.config_map
+            .get(canonical_path.parent().unwrap())
+            .unwrap()
+            .get_config(canonical_path)
+    }
+}
+
+impl AssetConfigTreeNode {
+    fn load(
+        parent: Option<Arc<AssetConfigTreeNode>>,
+        dir: &Path,
+        configs: &mut HashMap<PathBuf, Arc<AssetConfigTreeNode>>,
+    ) -> anyhow::Result<()> {
+        let mut rules = vec![];
+        if let Ok(file) = File::open(dir.join(ASSETS_CONFIG_FILENAME)) {
+            let reader = BufReader::new(file);
+            CURRENTLY_PROCESSED_ASSETS_DIRECTORY.with(|book| {
+                let x = book.borrow_mut();
+                let mut path = RefMut::map(x, |p| p);
+                *path = dir.to_path_buf();
+            });
+
+            match serde_json::from_reader(reader) {
+                Ok(mut v) => rules.append(&mut v),
+                Err(e) => bail!("{:?}: {:?}", dir, e),
+            }
+        }
+
+        let config_tree = Self { parent, rules };
+        let parent_ref = Arc::new(config_tree);
+        configs.insert(dir.to_path_buf(), parent_ref.clone());
+        for f in std::fs::read_dir(&dir)
+            .unwrap()
+            .filter_map(|x| x.ok())
+            .filter(|x| x.file_type().unwrap().is_dir())
+        {
+            Self::load(Some(parent_ref.clone()), &f.path(), configs)?;
+        }
+        Ok(())
+    }
+
+    fn get_config(&self, canonical_path: &Path) -> AssetConfig {
+        let base_config = match &self.parent {
+            Some(parent) => parent.get_config(&canonical_path),
+            None => AssetConfig::default(),
+        };
+        self.rules
+            .iter()
+            .cloned()
```

Comment:
yes and no. In that particular spot, cloning was not good, because it was called before both `.filter` and `.map`, whereas my intention was to call it just before `.map`. I moved it downwards in the call stack to a more specific spot.

It is necessary to `clone`, because `AssetConfig::merge` requires to have owned data when we're trying to merge `self` with `other`